src/lib/expression/abstract_expression.cpp

```diff
@@ -0,0 +1,56 @@
+#include "abstract_expression.hpp"
+
+#include <queue>
+#include <string>
+
+#include "boost/functional/hash.hpp"
+#include "expression_utils.hpp"
+#include "utils/assert.hpp"
+
+using namespace std::string_literals;
+
+namespace opossum {
+
+AbstractExpression::AbstractExpression(const ExpressionType type,
+                                       const std::vector<std::shared_ptr<AbstractExpression>>& arguments)
+    : type(type), arguments(arguments) {}
+
+bool AbstractExpression::requires_calculation() const { return !arguments.empty(); }
+
+bool AbstractExpression::is_nullable() const {
+  return !std::all_of(arguments.begin(), arguments.end(),
+                      [](const auto& expression) { return !expression->is_nullable(); });
+}
+
+bool AbstractExpression::deep_equals(const AbstractExpression& other) const {
```

Comment:
`deep_equals` is more obvious to do a deep traversal of a tree than operator==, which I would expect to be a shallow equal.

We wouldn't be able to use vector::operator== because that would just compare the pointers.

Reply:
as discussed: `op==`